packages/babel-parser/src/parser/statement.js

```diff
@@ -818,6 +819,9 @@ export default class StatementParser extends ExpressionParser {
     if (createNewLexicalScope) {
       this.scope.exit();
     }
+    if (afterDirectivesParse) {
+      afterDirectivesParse.call();
```

Comment:
By calling it here, you are calling it after that the whole block has been parsed, not just the directives. You should call it inside `parseBlockOrModuleBlockBody`, when setting `parsedNonDirective` for the first time.

https://github.com/babel/babel/blob/ad29d91be6fb6969e3a15eac4037684ec7935800/packages/babel-parser/src/parser/statement.js#L885-L888

We also need to be careful: if we never reach the linked code (in case there are only directives and nothing else), we also need to call the function.

Also, `.call()` without any argument isn't needed, you can just use `afterDirectivesParse`.

---

Note that this would make the function be called after we parse a non-directive statement, but I think that it's ok since I couldn't think of a case which could be broken by this change.

This would still work, which is the only affected case:
```js
function f(a) {
  "use strict";
  let a;
}
```
because before registering the `a` parameter we have already registered `let a` to the scope, and thus we can catch duplicates.

Reply:
This means we need to pass `afterDirectivesParse` all the way down through `parseBlock`, `parseBlockBody`, and finally `parseBlockOrModuleBlockBody`, correct?

I might be having trouble with calling the callback when only directives are parsed. I'll go ahead and push the changes that I have.

Reply:
If after the loop it `parsedNonDirective` is still true, you'll need to call the callback.

Reply:
Ah that may be where I was having trouble. I was doing the following:
https://github.com/babel/babel/blob/355ee3ee2b62ac020c947e35117f6faffd1915e0/packages/babel-parser/src/parser/statement.js#L900-L902
which means I was calling the callback if `parsedNonDirective` was false.

Reply:
Oh I was wrong, your code is correct :man_facepalming: